src/System.Slices/System/Buffers/ReferenceCounter.cs

```diff
@@ -85,129 +85,86 @@ private static ObjectTable AddThreadLocalTable()
         }
     }
 
-    // This datastructure holds a collection of object to reference count mappings
+    // This datastructure holds a collection of objects to reference count mappings
+    // Uses repetition of entries to represent the count.
     sealed class ObjectTable
     {
         // if you change this constant, update ResizingObjectTableWorks test.
         const int DefaultTableCapacity = 16;
 
-        ObjectReferences[] _items;
-        int _count;
-
-        struct ObjectReferences
-        {
-            public object Obj;
-            public uint ReferenceCount;
-
-            public override string ToString()
-            {
-                return ReferenceCount.ToString();
-            }
-        }
+        object[] _items;
+        int _first;
         
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal void Increment(object obj)
+        public ObjectTable()
         {
-            if (_items == null)
-            {
-                _items = new ObjectReferences[DefaultTableCapacity];
-            }
-            var index = FindExistingOrNewIndex(obj);
-            _items[index].ReferenceCount++;
+            _items = new object[DefaultTableCapacity];
+            _first = _items.Length;
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal void Decrement(object obj)
+        internal void Increment(object obj)
         {
-            var index = FindExistingIndex(obj);
-            if(index == -1) {
-                ThrowCountNotPositive();
+            if(_first == 0)
+            {
+                // No space left, we must grow the table.
+                var oldLength = _items.Length;
+                var larger = new object[oldLength << 1];
+                _items.CopyTo(larger, oldLength);
+                // The copy must occur before the table is published.
+                Volatile.Write(ref _items, larger);
+                // _items must be updated before the value of _first is changed
+                // If this is not the case, then we could effectively make the
+                // table empty temporarily.
+                _items[oldLength - 1] = obj;
+                Volatile.Write(ref _first, oldLength - 1);
```

Comment:
I believe it is, if the write at 119 was to move above 115, then the effect would be make the array appear to contain just one element, until the update to _items is propogated.  Now, I am not sure if it is something that any implementations do, but I thought it was allowed by the ECMA spec.